include/boost/uuid/detail/random_provider.hpp

```diff
@@ -0,0 +1,75 @@
+//
+// Copyright (c) 2017 James E. King III
+//
+// Distributed under the Boost Software License, Version 1.0.
+// (See accompanying file LICENSE_1_0.txt or copy at
+//   http://www.boost.org/LICENCE_1_0.txt)
+//
+// Platform-specific random entropy provider
+//
+
+#ifndef BOOST_UUID_DETAIL_RANDOM_PROVIDER_HPP
+#define BOOST_UUID_DETAIL_RANDOM_PROVIDER_HPP
+
+#include <boost/core/noncopyable.hpp>
+#include <boost/cstdint.hpp>
+#include <boost/limits.hpp>
+#include <boost/static_assert.hpp>
+#include <boost/type_traits/is_integral.hpp>
+#include <boost/type_traits/is_unsigned.hpp>
+#include <boost/uuid/entropy_error.hpp>
+#include <climits>
+#include <iterator>
+
+// Detection of the platform is separated from inclusion of the correct
+// header to facilitate mock testing of the provider implementations.
+
+#include <boost/uuid/detail/random_provider_detect_platform.hpp>
+#include <boost/uuid/detail/random_provider_include_platform.hpp>
+
+
+namespace boost {
+namespace uuids {
+namespace detail {
+
+//! \brief Contains code common to all random_provider implementations.
+//! \note  random_provider_base is required to provide this method:
+//!        void get_random_bytes(void *buf, size_t siz);
+//! \note  noncopyable because of some base implementations so
+//!        this makes it uniform across platforms to avoid any  
+//!        porting surprises
+class random_provider
+    : public detail::random_provider_base,
+      public noncopyable
+{
+public:
+    //! Leverage the provider as a SeedSeq for
+    //! PseudoRandomNumberGeneration seeing.
+    //! \note: See Boost.Random documentation for more details
+    template<class Iter>
+    void generate(Iter first, Iter last)
+    {
+        typedef typename std::iterator_traits<Iter>::value_type value_type;
+        BOOST_STATIC_ASSERT(is_integral<value_type>::value);
+        BOOST_STATIC_ASSERT(is_unsigned<value_type>::value);
+        BOOST_STATIC_ASSERT(sizeof(value_type) * CHAR_BIT >= 32);
+
+        for (; first != last; ++first)
+        {
+            get_random_bytes(&*first, sizeof(*first));
+            *first &= (std::numeric_limits<boost::uint32_t>::max)();
```

Comment:
`uint32_t` is not required to exist, and we already know its `max`: `0xFFFFFFFF`. So just use it.

Reply:
`<boost/cstdint.hpp>` guarantees a proper definition.  Isn't `std::numeric_limits<T>::max` a compile-time constant?  It shouldn't matter.

Reply:
uint32_t only exists when the platform has a 32 bit type.

Why would anyone prefer to write `(std::numeric_limits<boost::uint32_t>::max)();` instead of `0xFFFFFFFF` is beyond me.

Reply:
0xFFFFFFFF. :-)